instrument/runtime.go

```diff
@@ -0,0 +1,182 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package instrument
+
+import (
+	"os"
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/shirou/gopsutil/process"
+	"github.com/uber-go/tally"
+)
+
+// StartReportingExtendedMetrics starts reporting extended metrics.
+func StartReportingExtendedMetrics(
+	scope tally.Scope,
+	reportInterval time.Duration,
+) (*ExtendedMetricsReporter, error) {
+	reporter, err := NewExtendedMetricsReporter(scope, reportInterval)
+	if err != nil {
+		return nil, err
+	}
+	reporter.Start()
+	return reporter, nil
+}
+
+type runtimeMetrics struct {
+	NumGoRoutines   tally.Gauge
+	GoMaxProcs      tally.Gauge
+	MemoryAllocated tally.Gauge
+	MemoryHeap      tally.Gauge
+	MemoryHeapIdle  tally.Gauge
+	MemoryHeapInuse tally.Gauge
+	MemoryStack     tally.Gauge
+	NumGC           tally.Counter
+	GcPauseMs       tally.Timer
+	lastNumGC       uint32
+}
+
+type processMetrics struct {
+	NumFds      tally.Gauge
+	NumFdErrors tally.Counter
+	process     *process.Process
+}
+
+// ExtendedMetricsReporter reports an extended set of metrics.
+type ExtendedMetricsReporter struct {
+	reportInterval time.Duration
+	runtime        runtimeMetrics
+	process        processMetrics
+	started        bool
+	quit           chan struct{}
+}
+
+// NewExtendedMetricsReporter creates a new extended metrics reporter.
+func NewExtendedMetricsReporter(
+	scope tally.Scope,
+	reportInterval time.Duration,
+) (*ExtendedMetricsReporter, error) {
+	pid := os.Getpid()
+	process, err := process.NewProcess(int32(pid))
+	if err != nil {
+		return nil, err
+	}
+
+	var memstats runtime.MemStats
+	runtime.ReadMemStats(&memstats)
+
+	runtimeScope := scope.SubScope("runtime")
+	memoryScope := runtimeScope.SubScope("memory")
+	processScope := scope.SubScope("process")
+
+	return &ExtendedMetricsReporter{
+		reportInterval: reportInterval,
+		runtime: runtimeMetrics{
+			NumGoRoutines:   runtimeScope.Gauge("num-goroutines"),
+			GoMaxProcs:      runtimeScope.Gauge("gomaxprocs"),
+			MemoryAllocated: memoryScope.Gauge("allocated"),
+			MemoryHeap:      memoryScope.Gauge("heap"),
+			MemoryHeapIdle:  memoryScope.Gauge("heapidle"),
+			MemoryHeapInuse: memoryScope.Gauge("heapinuse"),
+			MemoryStack:     memoryScope.Gauge("stack"),
+			NumGC:           memoryScope.Counter("num-gc"),
+			GcPauseMs:       memoryScope.Timer("gc-pause-ms"),
+			lastNumGC:       memstats.NumGC,
+		},
+		process: processMetrics{
+			NumFds:      processScope.Gauge("num-fds"),
+			NumFdErrors: processScope.Counter("num-fd-errors"),
+			process:     process,
+		},
+		started: false,
+		quit:    make(chan struct{}),
+	}, nil
+}
+
+// Start starts the reporter thread that periodically emits metrics.
+func (r *ExtendedMetricsReporter) Start() {
+	if r.started {
```

Comment:
nit: We may want to consider using a mutex here to protect `r.started` since there's still a possibility, albeit very small, of `Start` being called concurrently and the two report goroutines being kicked off.